### ISSUE TITLE
More flexible configuration inheritance

### ISSUE THREAD

**Issue by lptr** [gradle/gradle-profiler#163]
In many cases scenarios look a lot alike, differing in only a few values. The currently used HOCON configuration allows inheriting properties from one or more parents, and this works well when properties are atomic.

However, in many cases we need to configure array-style properties. It is not currently possible to add more options to inherited `gradle-args`.

It would be great if we could do something with mixins instead of inheritance, too. Here's an idea of how to make this work with Groovy's `ConfigSlurper`:

https://gist.github.com/lptr/2c9a9ae7c7d7296c7628518380b2541b

Groovy is pretty well-suited for this anyway.

**Comment by eskatos** [comment]
Doesn't this work already?

```
gradle-args = $gradle-args ["more", "args"]
```
https://github.com/lightbend/config/blob/master/HOCON.md#array-and-object-concatenation

**Comment by lptr** [comment]
Unfortunately no.

**Comment by lptr** [comment]
It would be nice to supply further mixins from the command-line (so you can turn caching off or on for example for all tested scenarios).

**Comment by lptr** [comment]
@wolf IIRC you wanted to improve the "programmatic scenario API". Could we do it together with this step, and introduce something where we could use, say, a Groovy DSL to configure scenarios instead of HOCON?

**Comment by wolfs** [comment]
Yes, exposing a better programmatic API which then can be configured by a Groovy DSL should work well.

**Comment by Mishkun** [comment]
To add an element to an array you can use this notation
```hocon
clean_build_without_cache {
  # configure scenario
}

clean_build_without_cache_buildeval = ${clean_build_without_cache}
clean_build_without_cache_buildeval.gradle-args = ${clean_build_without_cache.gradle-args} ["-Pbuildeval"]
```
